### ISSUE TITLE
Not compatible with 4.5 internal API

### ISSUE THREAD

**Issue by kungfoo** [johnrengelman/gradle-processes#9]
The internal API was extended with the `Executor` to be used.

Please see here:
https://github.com/gradle/gradle/blob/v4.5.0/subprojects/core/src/main/java/org/gradle/process/internal/JavaExecHandleBuilder.java

This means that any project or plugin using this plugin is going to fail when creating tasks on 4.5.

**Comment by kungfoo** [comment]
I've created a small project that shows the error:

https://github.com/kungfoo/gradle-processes-bug-9

**Comment by kungfoo** [comment]
And here is the output when it fails: https://travis-ci.org/kungfoo/gradle-processes-bug-9

**Comment by guspower** [comment]
I've just taken [a quick pass at upgrading the plugin](https://github.com/guspower/gradle-processes) to work with gradle 4.5. I've commented in #8 to this effect.

**Comment by kungfoo** [comment]
Yeah, stumbled across the same issue. I thought that 2.11 being stone-age, the bug report is probably also ages old :)

**Comment by guspower** [comment]
Heh yeah I would've done the same :) 

_looks for stoneage emoticon_

**Comment by kungfoo** [comment]
Maybe even older: 🦕 🦖 🤣

**Comment by guspower** [comment]
```
               __
              / _)
     _/\/\/\_/ /
   _|         /
 _|  (  | (  |
/__.-'|_|--|_|  
```

**Comment by kungfoo** [comment]
What's funny that at least the sample project I created, appears to work
with 3.5 for example, which (obviously) is quite a bit newer than 2.11

**Comment by ldaley** [comment]
This is blocking Ratpack's build from moving to Gradle 4.5 also.

**Comment by DJViking** [comment]
This plugin is the only one I have found that would behave like the process-exec-maven-plugin.
It is exactly what I am looking for to allow running a ruby server mock prior to running the tests. The normal Exec will not spawn a new process and thus halting the entire gradle build.

This plugin seems dead since the creator hasn't done any maintenance for 4 years.

The task name Fork and JavaFork (forked processes): Why not instead something like ExecFork or ExecProcess (to indicate it is an extension of Exec).

**Comment by kungfoo** [comment]
@guspower I took the liberty of using your changes for inspiration but to also remove some cruft from the build files of the plugins, so that it can be built anywhere.
I'll try to also add a travis build file and then maybe fork under a different name or adopt it.

It would be really cool if @johnrengelman could chime in on this issue since it's blocking a lot of things.

**Comment by DJViking** [comment]
If @johnrengelman is not willing to work further on this plugin, then we need to fork it under a different name where it can be continued to be developed.

**Comment by kungfoo** [comment]
I've already started working on either a patch or a fork under a different
name.

I'll publish the changes tomorrow to a forked repo and if there is no reply
within the next days I'd say it seems fair to fork and rename while giving
credit and keeping a link back to the old repo.

**Comment by johnrengelman** [comment]
Hey everyone. Thanks the leg work. If someone wants to submit a PR, I’ll get a release out this week.

**Comment by kungfoo** [comment]
Do you mind if I remove some of the gradle build files that keep it from building in other places, such as references to Bloom Health Codenarc configs?
I'll also add a travis build file, so that one can see if a specific version builds somewhere else on a clean-slate system.

**Comment by kungfoo** [comment]
While drawing inspiration from the heavy lifting of @guspower I created this feature branch https://github.com/kungfoo/gradle-processes/tree/feature/update-to-gradle-4.5-api

If refactored some of the tests along the way, and the plugin now should build on travis.

I'll have to check back after lunch. :)

**Comment by johnrengelman** [comment]
Thanks for cleaning up the build.

**Comment by kungfoo** [comment]
Of course you would have to re-add the bintray publishing part, but I thought that if there is going to be a new version, it will probably be published to https://plugins.gradle.org/

Let's see if we can get the build running on travis.

**Comment by kungfoo** [comment]
Okay, after fixing the download of the gradle distribution zip via https on ubuntu trusty, the build works on the following jdks on travis:

- oracle jdk 8
- open-jdk 8
- open-jdk 7

I think that's enough for now. :)

You can check the last builds that ran for the branch here: https://travis-ci.org/kungfoo/gradle-processes/builds/338479995

**Comment by kungfoo** [comment]
Maybe one thing that could be added is that the plugin checks the gradle version when it is added to the project and informs developers about the fact that the new version incompatible with older gradle versions instead of failing with a stacktrace. Let's see if there's an API for that.

**Comment by kungfoo** [comment]
I also added the check for the gradle version and a test that ensures it's actually displaying a nice error message.

One could also conditionally pick a different implementation and have it dynamically compiled (see here: https://discuss.gradle.org/t/specify-minimum-gradle-version-for-a-plugin/24536/3), but IMHO that seems like overkill. What do you think @johnrengelman ?

**Comment by DJViking** [comment]
New version published to https://plugins.gradle.org/ sounds like a good idea.

**Comment by johnrengelman** [comment]
I think just requiring the current version of Gradle is the best plan. I’ll work in getting the changes in and published tonight.

**Comment by kungfoo** [comment]
I also created a new PR #11 so you can easily publish a new version to plugins.gradle.org. @johnrengelman : Would be nice if you could merge and publish 0.4.0, since our build is still broken on gradle > 4.5...

**Comment by johnrengelman** [comment]
I've got everything set up for publishing, I'm waiting for Gradle to relink the plugin to my account since it was originally published on Bintray. I'll post here once the plugin has been released.

**Comment by johnrengelman** [comment]
I've published version 0.4.0 to the plugin portal. Please give it a try

**Comment by kungfoo** [comment]
Will do as soon as i can, I'll also create pull requests for the downstream
plugins that are failing our builds.

Thanks!

**Comment by hmortensen** [comment]
gradle-processes 0.4.0 resolved my gradle-compass problem with gradle 4.5.
Thanks.

**Comment by KrzysztofKowalczyk** [comment]
There is a bug in the new code, on JavaFork, the following should return `this` instead

```
    @Override
    JavaFork setArgs(List<String> list) {
        return forkAction.setArgs(list)
    }
```

**Comment by johnrengelman** [comment]
I published 0.4.1 to fix the incorrect return.

**Comment by nperumal** [comment]
The 0.4.1 plugin throws the following with gradle 4.5.1 and 4.2.1 on [apply plugin: 'com.github.johnrengelman.processes'] 

Is the id 'com.github.johnrengelman.processes' correct for 0.4.1?

> Failed to apply plugin [id 'com.github.johnrengelman.processes']
   > This version of the plugin is incompatible with gradle < 4.5! Use version 0.3.0 for now.
